File: PROGRAMMERS/LEVEL_02/Q12945.cpp
//https://programmers.co.kr/learn/courses/30/lessons/12945
//피보나치 수
#include <iostream>

#include <string>
#include <vector>

using namespace std;

int solution(int n) {

    const int MOV = 1234567;
    const int MAX = 100000;
    int fibo[MAX + 1] = { 0, };

    fibo[1] = 1;
    for (int i = 2; i <= n; ++i)
    {
        fibo[i] = (fibo[i - 1] + fibo[i - 2]) % MOV;
    }

    return fibo[n];
}

int main()
{
    int n;
    cin >> n;
    cout << solution(n);

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q12945.java
// https://programmers.co.kr/learn/courses/30/lessons/12945?language=java
// 피보나치 수
package PROGRAMMERS.LEVEL_02;

import java.util.Arrays;

public class Q12945 {

    public static void main(String[] args) {
        /*
        n	return
        3	2
        5	5
         */
        int n = 3;
        System.out.println(solution(n));
    }

    public static int solution(int n) {

        int[] memo = new int[n + 1];
        Arrays.fill(memo, -1);
        memo[0] = 0;
        memo[1] = 1;

        return fibonacci(n, memo);
    }

    public static int fibonacci(int n, int[] memo) {
        if (memo[n] >= 0) {
            return memo[n];
        }
        return memo[n] = (fibonacci(n - 1, memo) + fibonacci(n - 2, memo)) % 1234567;
    }
}
